### ISSUE TITLE
SFTTrainer should not set tokenizer.pad_token_id = tokenizer.eos_token_id

### ISSUE THREAD

**Issue by kdmcdrm** [huggingface/trl#1384]
I would like to suggest that SFTTrainer should not set tokenizer.pad_token_id = tokenizer.eos_token_id when the tokenizer does not have a set pad_token_id, as it currently does on line 219 of sft_trainer.py (version 0.7.11) as the result is that the model is fine-tuned on samples without an eos token, and therefore generates too much text (rambles).

I'd suggest that the alternate approach of defining a new pad token that is suggested in [this thread](https://github.com/huggingface/transformers/issues/22794) should be applied instead. For example

``` python
tokenizer.add_special_tokens({'pad_token': '[PAD]'})
model.resize_token_embeddings(len(tokenizer))
# Tell the model about the pad token id, not sure if this is needed
model.config.pad_token_id = tokenizer.pad_token_id
````

I'm interested to hear if I've missed something, but setting pad_token_id to eos_token_id doesn't seem to make sense unless (maybe) you also ensure packing=True?

**Comment by fancyerii** [comment]
I haven't tested whether fast tokenizer still has the bug as in [this issue](https://github.com/huggingface/transformers/issues/22794). But it's merged to master since 4/24/2023. So if don't use old transformers, there is no problem.

As for SFTTrainer, I read/debug the code again and am pretty sure that at the end of  every question/answer pair, a concat_token_id token is appended. You can check the code [here](https://github.com/huggingface/trl/blob/v0.7.11/trl/trainer/utils.py#L465). 
When I debug use the default script setting. concat_token_id is set to 2(</eos>).

![image](https://github.com/huggingface/trl/assets/5372812/3d0f20a8-0e62-4333-93b6-fa45656996e9)

**Comment by kdmcdrm** [comment]
Thanks for looking into this. 

> If you want to train each qa pair as an example and pad to a fixed length or use datacollator to pad to longest length on the fly, you can use a new <pad> token.

Agreed, I think it would have worked if I was using a FixedLengthDataset due to the code you show above, but I was  using a `formatting_func` and packing=False, as per the example in the documentation - https://huggingface.co/docs/trl/en/sft_trainer#customize-your-prompts-using-packed-dataset. 

> But if it's decoder only model like llama and padding_side is right, I think use <eos> is also ok. Because in pretraining, <pad> is not used. Meta use methods similar to ConstantLengthDataset to pretrain, so they do not use <pad>.
But if you use <eos> as pad, make sure padded label is set to -100 to skip loss computation.

This is incorrect for the reason I provide above, I observed that all eos_token were ignored (set to -100) during training, and therefore the model rambled. To fix it I had to add a pad token to the model and tokenizer, and manually affix an eos token to the end of each example.

Overall, I'd still suggest that the library not automatically set pad_token_id=eos_token_id unless it ensures that the eos token will be included in each sample.
